pkg/typesystem/typesystem.go

```diff
@@ -560,6 +567,67 @@ func (t *TypeSystem) IsDirectlyRelated(target *openfgav1.RelationReference, sour
 	return false, nil
 }
 
+func (t *TypeSystem) TTUCanFastPathLevel(objectType, relation, userType string, ttu *openfgav1.TupleToUserset, level int) bool {
+	if t.authzWeightedGraph == nil {
+		return false
+	}
+	objRel := tuple.ToObjectRelationString(objectType, relation)
+	tuplesetRelationKey := tuple.ToObjectRelationString(objectType, ttu.GetTupleset().GetRelation())
+	computedRelation := ttu.GetComputedUserset().GetRelation()
+	node, ok := t.authzWeightedGraph.GetNodesById(objRel)
+	if !ok {
+		return false
+	}
+
+	w, ok := node.GetWeight(userType)
+	if !ok {
+		return false
+	}
+
+	if w == level {
+		return true
+	}
+
+	edges, ok := t.authzWeightedGraph.GetEdgesByNode(node)
+	if !ok {
+		return false
+	}
+
+	ttuEdges := make([]*graph.WeightedAuthorizationModelEdge, 0)
+
+	// find all TTU edges  with valid weight
```

Comment:
I'm going to defer to @yissellokta to give a more detailed explanation.

Reply:
yes, it is the maximum weight, so you can have a node with weight 3 but that does not mean that the TTU edge you are evaluating is weight 3, that TTU edge can have weight 2 but there is another TTU edge or userset edge, etc that can have weight 3. When the weighted graph gets passed and used instead of the type system, all this evaluation will be simpler because you will be passing directly the TTU edge while traversing the graph, instead to what we have right now